### PR TITLE
[Merged by Bors] - feat(SetTheory/Game/Birthday): games with bounded birthday are a small set

### PR DESCRIPTION
**Pull request by vihdzp**
---

- [x] depends on: #15716

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/SetTheory/Game/Birthday.lean [hunk 1]
```diff
@@ -270,6 +270,52 @@ theorem birthday_sub_le (x y : Game) : (x - y).birthday ≤ x.birthday ♯ y.bir
...
+      · refine ⟨(equivShrink {y // (g y).2}) ⟨⟨⟦y.moveRight i⟧, ?_⟩, i, rfl⟩, ?_⟩
+        · suffices ∃ b ≤ a, birthday ⟦y.moveRight i⟧ < b from by simpa [S] using this
+          refine ⟨_, le_of_eq hx, ?_⟩
+          rw [← hy']
+          exact (birthday_quot_le_pGameBirthday _).trans_lt (PGame.birthday_moveRight_lt i)
+        · rw [PGame.moveRight_mk, Equiv.symm_apply_apply]
+          exact Quotient.mk_out _
+  · convert H
+    change birthday _ < o ↔ ∃ a, _
+    simpa using lt_limit ho
```
Comment (b-mehta) [inline Mathlib/SetTheory/Game/Birthday.lean:317]:
```suggestion
  induction o using Ordinal.induction with | h o IH =>
  let S := ⋃ a ∈ Set.Iio o, {x : Game.{u} | birthday x < a}
  let H : Small.{u} S := @small_biUnion _ _ _ _ _ IH
  obtain rfl | ⟨a, rfl⟩ | ho := zero_or_succ_or_limit o
  · simp_rw [Ordinal.not_lt_zero]
    exact small_empty
  · simp_rw [Order.lt_succ_iff, le_iff_lt_or_eq]
    convert small_union.{u} {x | birthday x < a} {x | birthday x = a}
    · exact IH _ (Order.lt_succ a)
    · let f (g₁ g₂ : Set S) : Game := ⟦PGame.mk _ _
        (fun x ↦ ((equivShrink g₁).symm x).1.1.out) (fun x ↦ ((equivShrink g₂).symm x).1.1.out)⟧
      suffices {x | x.birthday = a} ⊆ Set.image2 f Set.univ Set.univ from small_subset this
      rintro x rfl
      obtain ⟨y, rfl, hy'⟩ := birthday_eq_pGameBirthday x
      let g₁ : Set S := {z | ∃ i, ⟦y.moveLeft i⟧ = z.1}
      let g₂ : Set S := {z | ∃ i, ⟦y.moveRight i⟧ = z.1}
      refine ⟨g₁, by simp, g₂, by simp, ?_⟩
      apply PGame.game_eq <| PGame.Equiv.of_exists _ _ _ _ <;>
      intro i
      · obtain ⟨j, hj⟩ := ((equivShrink g₁).symm i).2
        exact ⟨j, by simp [PGame.equiv_iff_game_eq, hj]⟩
      · obtain ⟨j, hj⟩ := ((equivShrink g₂).symm i).2
        exact ⟨j, by simp [PGame.equiv_iff_game_eq, hj]⟩
      · refine ⟨equivShrink g₁ ⟨⟨⟦y.moveLeft i⟧, ?_⟩, i, rfl⟩, by simp [Quotient.mk_out]⟩
        suffices ∃ b ≤ y.birthday, birthday ⟦y.moveLeft i⟧ < b by simpa [S, hy'] using this
        refine ⟨_, le_rfl, ?_⟩
        exact (birthday_quot_le_pGameBirthday _).trans_lt (PGame.birthday_moveLeft_lt i)
      · refine ⟨equivShrink g₂ ⟨⟨⟦y.moveRight i⟧, ?_⟩, i, rfl⟩, by simp [Quotient.mk_out]⟩
        suffices ∃ b ≤ y.birthday, birthday ⟦y.moveRight i⟧ < b by simpa [S, hy'] using this
        refine ⟨_, le_rfl, ?_⟩
        exact (birthday_quot_le_pGameBirthday _).trans_lt (PGame.birthday_moveRight_lt i)
  · convert H
    change birthday _ < o ↔ ∃ a, _
    simpa using lt_limit ho
```
Here's a golf, if you'd like

**Review by b-mehta — APPROVED** [review]

File: Mathlib/SetTheory/Game/PGame.lean [hunk 2]
```diff
@@ -822,12 +822,32 @@ theorem equiv_congr_right {x₁ x₂ : PGame} : (x₁ ≈ x₂) ↔ ∀ y₁, (x
...
+    exact Or.inr ⟨j, Equiv.le hj⟩
+  · obtain ⟨j, hj⟩ := hl₂ i
+    exact Or.inl ⟨j, Equiv.ge hj⟩
+  · obtain ⟨j, hj⟩ := hr₁ i
+    exact Or.inr ⟨j, Equiv.ge hj⟩
+
 theorem Equiv.of_equiv {x y : PGame} (L : x.LeftMoves ≃ y.LeftMoves)
     (R : x.RightMoves ≃ y.RightMoves) (hl : ∀ i, x.moveLeft i ≈ y.moveLeft (L i))
     (hr : ∀ j, x.moveRight j ≈ y.moveRight (R j)) : x ≈ y := by
-  constructor <;> rw [le_def]
```
[the remark this answers is not in the document]
Comment (vihdzp) [reply Mathlib/SetTheory/Game/PGame.lean:828]: I thought it'd make sense for this to be proved in terms of `Equiv.of_exists`. The proof is pretty much the same besides that.

[the remark this answers is not in the document]
Comment (vihdzp) [reply Mathlib/SetTheory/Game/PGame.lean:833]: I've golfed this theorem a bit, the correspondence between the old and new proof should now be much more obvious.

File: Mathlib/SetTheory/Game/Birthday.lean (hunk 1 above)
Comment (vihdzp) [reply Mathlib/SetTheory/Game/Birthday.lean]: Thanks! I took the opportunity to golf this even further ;)

**Comment by b-mehta** [comment]
Thanks!

bors merge

**Pull request closed without merging**